File: src/main/scala/laoflch/debezium/connector/informix/InformixChangeRecordEmitter.scala
/*
 * Copyright Debezium Authors.
 *
 * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 */
package laoflch.debezium.connector.informix.integrtest

import com.informix.jdbc.IfmxReadableType
import io.debezium.data.Envelope
import io.debezium.data.Envelope.Operation
import io.debezium.pipeline.spi.OffsetContext
import io.debezium.relational.RelationalChangeRecordEmitter
import io.debezium.util.Clock

import scala.jdk.CollectionConverters


/**
 * Emits change data based on a single (or two in case of updates) CDC data row(s).
 *
 * @author Jiri Pechanec
 */
object InformixChangeRecordEmitter {
  val OP_DELETE = 0x01
  val OP_INSERT = 0x02
  val OP_UPDATE = 0x03
  //val OP_UPDATE_AFTER = 0x04
  val OP_TRUNCATE = 0x05

  /**
   * convert columns data from map[String,IfmxReadableType] to array[AnyRef].
   * debezium can't convert the IfmxReadableType object to kafka direct,so use map[AnyRef](x=>x.toObject) to extract the jave type value
   * from IfmxReadableType and pass to debezium for kafka
   * @param data the data from informix cdc map[String,IfmxReadableType].
   * @author laoflch luo
   */
  def convertIfxData2Array(data: java.util.Map[String, IfmxReadableType]): Array[AnyRef] = {
    if (data == null) {return Array.emptyObjectArray}

    CollectionConverters.CollectionHasAsScala(data.values()).asScala.map[AnyRef](x=>x.toObject).toArray
    //extract value from IfmxReadableType and pass it to ValueConverter for kafka value

  }
}
class InformixChangeRecordEmitter(val offset: OffsetContext, val operation: Int, val data: Array[AnyRef], val dataNext: Array[AnyRef], val clock: Clock) extends RelationalChangeRecordEmitter(offset, clock) {
  override protected def getOperation: Envelope.Operation = {
    if (operation == InformixChangeRecordEmitter.OP_DELETE) return Operation.DELETE
    else if (operation == InformixChangeRecordEmitter.OP_INSERT) return Operation.CREATE
    else if (operation == InformixChangeRecordEmitter.OP_UPDATE) return Operation.UPDATE
    throw new IllegalArgumentException("Received event of unexpected command type: " + operation)
  }

  override protected def getOldColumnValues: Array[AnyRef] = getOperation match {
    case Operation.CREATE =>null
    case Operation.READ =>null
    case _ =>
      data
  }

  override protected def getNewColumnValues: Array[AnyRef] = getOperation match {
    case Operation.CREATE =>dataNext
    case Operation.READ =>
      data
    case Operation.UPDATE =>
      dataNext
    case _ =>
      null
  }


}

File: src/main/java/laoflch/debezium/connector/informix/InformixChangeRecordEmitter.java
/*
 * Copyright Debezium-Informix-Connector Authors.
 *
 * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 */

package laoflch.debezium.connector.informix;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.apache.kafka.connect.data.Field;
import org.apache.kafka.connect.data.Struct;

import com.informix.jdbc.IfmxReadableType;

import io.debezium.data.Envelope.Operation;
import io.debezium.pipeline.spi.OffsetContext;
import io.debezium.relational.RelationalChangeRecordEmitter;
import io.debezium.relational.TableSchema;
import io.debezium.util.Clock;

public class InformixChangeRecordEmitter extends RelationalChangeRecordEmitter {

    public static final int OP_DELETE = 1;
    public static final int OP_INSERT = 2;
    public static final int OP_UPDATE = 3;
    // public static final int OP_UPDATE_AFTER = 4;
    public static final int OP_TRUNCATE = 5;

    private final int operation;
    private final Object[] data;
    private final Object[] dataNext;

    public InformixChangeRecordEmitter(OffsetContext offset, int operation, Object[] data, Object[] dataNext, Clock clock) {
        super(offset, clock);

        this.operation = operation;
        this.data = data;
        this.dataNext = dataNext;
    }

    @Override
    protected Operation getOperation() {
        if (operation == OP_DELETE) {
            return Operation.DELETE;
        }
        else if (operation == OP_INSERT) {
            return Operation.CREATE;
        }
        else if (operation == OP_UPDATE) {
            return Operation.UPDATE;
        }
        else if (operation == OP_TRUNCATE) {
            return Operation.TRUNCATE;
        }
        throw new IllegalArgumentException("Received event of unexpected command type: " + operation);
    }

    @Override
    protected Object[] getOldColumnValues() {
        switch (getOperation()) {
            case CREATE:
            case READ:
                return null;
            default:
                return data;
        }
    }

    @Override
    protected Object[] getNewColumnValues() {
        switch (getOperation()) {
            case CREATE:
            case UPDATE:
                return dataNext;
            case READ:
                return data;
            default:
                return null;
        }
    }

    /**
     * Convert columns data from Map[String,IfmxReadableType] to Object[].
     * Debezium can't convert the IfmxReadableType object to kafka direct,so use map[AnyRef](x=>x.toObject) to extract the jave type value
     * from IfmxReadableType and pass to debezium for kafka
     *
     * @param data the data from informix cdc map[String,IfmxReadableType].
     *
     * @author Laoflch Luo, Xiaolin Zhang
     */
    public static Object[] convertIfxData2Array(Map<String, IfmxReadableType> data, TableSchema tableSchema) throws SQLException {
        if (data == null) {
            return new Object[0];
        }

        List<Object> list = new ArrayList<>();
        for (Field field : tableSchema.valueSchema().fields()) {
            IfmxReadableType ifmxReadableType = data.get(field.name());
            Object toObject = ifmxReadableType.toObject();
            list.add(toObject);
        }
        return list.toArray();
    }

    @Override
    protected void emitTruncateRecord(Receiver receiver, TableSchema tableSchema) throws InterruptedException {
        Struct envelope = tableSchema.getEnvelopeSchema().truncate(getOffset().getSourceInfo(), getClock().currentTimeAsInstant());
        receiver.changeRecord(tableSchema, Operation.TRUNCATE, null, envelope, getOffset(), null);
    }
}
